File: multi_agent/projects/im-mobile/lib/services/message_expiration_service.dart
/**
 * 消息过期服务
 */
import 'package:dio/dio.dart';
import '../models/message_expiration_rule.dart';

class MessageExpirationService {
  final Dio _dio;

  MessageExpirationService(this._dio);

  /// 创建会话级过期规则
  Future<MessageExpirationRule> createRule(String conversationId, ExpirationRuleRequest req) async {
    final resp = await _dio.post('/api/expiration/rules', data: {
      ...req.toJson(),
      'conversationId': conversationId,
    });
    return MessageExpirationRule.fromJson(resp.data);
  }

  /// 更新过期规则
  Future<MessageExpirationRule> updateRule(int ruleId, ExpirationRuleRequest req) async {
    final resp = await _dio.put('/api/expiration/rules/$ruleId', data: req.toJson());
    return MessageExpirationRule.fromJson(resp.data);
  }

  /// 获取用户所有规则
  Future<List<MessageExpirationRule>> getUserRules() async {
    final resp = await _dio.get('/api/expiration/rules');
    final List<dynamic> data = resp.data ?? [];
    return data.map((e) => MessageExpirationRule.fromJson(e)).toList();
  }

  /// 获取会话生效规则
  Future<MessageExpirationRule?> getEffectiveRule(String conversationId) async {
    final resp = await _dio.get('/api/expiration/rules/conversation/$conversationId');
    if (resp.data == null) return null;
    return MessageExpirationRule.fromJson(resp.data);
  }

  /// 获取全局默认规则
  Future<MessageExpirationRule?> getGlobalRule() async {
    final resp = await _dio.get('/api/expiration/rules/global');
    if (resp.data == null) return null;
    return MessageExpirationRule.fromJson(resp.data);
  }

  /// 删除规则
  Future<void> deleteRule(int ruleId) async {
    await _dio.delete('/api/expiration/rules/$ruleId');
  }

  /// 启用/禁用规则
  Future<MessageExpirationRule> toggleRule(int ruleId, bool enabled) async {
    final resp = await _dio.patch('/api/expiration/rules/$ruleId/toggle?enabled=$enabled');
    return MessageExpirationRule.fromJson(resp.data);
  }

  /// 获取消息剩余存活时间
  Future<int> getRemainingTime(int messageId) async {
    final resp = await _dio.get('/api/expiration/messages/$messageId/remaining');
    return resp.data['remainingSeconds'] ?? -1;
  }

  /// 记录消息被阅读（启动阅后即焚计时）
  Future<void> recordRead(int messageId) async {
    await _dio.post('/api/expiration/messages/$messageId/read');
  }
}

File: im-desktop/src/services/message-expiration-service.ts
/**
 * 消息过期服务
 */
import { apiClient } from './api-client';
import type { ExpirationRule, ExpirationRuleRequest } from '../types/message-expiration';

export class MessageExpirationService {

  /**
   * 创建会话级过期规则
   */
  async createRule(conversationId: string, request: ExpirationRuleRequest): Promise<ExpirationRule> {
    return apiClient.post<ExpirationRule>('/api/expiration/rules', {
      ...request,
      conversationId,
    });
  }

  /**
   * 更新过期规则
   */
  async updateRule(ruleId: number, request: ExpirationRuleRequest): Promise<ExpirationRule> {
    return apiClient.put<ExpirationRule>(`/api/expiration/rules/${ruleId}`, request);
  }

  /**
   * 获取用户所有规则
   */
  async getUserRules(): Promise<ExpirationRule[]> {
    return apiClient.get<ExpirationRule[]>('/api/expiration/rules');
  }

  /**
   * 获取会话生效规则
   */
  async getEffectiveRule(conversationId: string): Promise<ExpirationRule | null> {
    return apiClient.get<ExpirationRule>(`/api/expiration/rules/conversation/${conversationId}`);
  }

  /**
   * 获取全局默认规则
   */
  async getGlobalRule(): Promise<ExpirationRule | null> {
    return apiClient.get<ExpirationRule>('/api/expiration/rules/global');
  }

  /**
   * 删除规则
   */
  async deleteRule(ruleId: number): Promise<void> {
    return apiClient.delete(`/api/expiration/rules/${ruleId}`);
  }

  /**
   * 启用/禁用规则
   */
  async toggleRule(ruleId: number, enabled: boolean): Promise<ExpirationRule> {
    return apiClient.patch<ExpirationRule>(`/api/expiration/rules/${ruleId}/toggle?enabled=${enabled}`);
  }

  /**
   * 获取消息剩余存活时间
   */
  async getRemainingTime(messageId: number): Promise<number> {
    const resp = await apiClient.get<{ messageId: number; remainingSeconds: number }>(
      `/api/expiration/messages/${messageId}/remaining`
    );
    return resp.remainingSeconds ?? -1;
  }

  /**
   * 记录消息被阅读（启动阅后即焚计时）
   */
  async recordRead(messageId: number): Promise<void> {
    return apiClient.post(`/api/expiration/messages/${messageId}/read`);
  }

  /**
   * 格式化过期时长为可读字符串
   */
  static formatExpiration(seconds: number): string {
    if (seconds < 60) return `${seconds}秒`;
    if (seconds < 3600) return `${Math.floor(seconds / 60)}分钟`;
    if (seconds < 86400) return `${Math.floor(seconds / 3600)}小时`;
    return `${Math.floor(seconds / 86400)}天`;
  }

  /**
   * 格式化相对时间
   */
  static formatRelative(seconds: number | null): string {
    if (!seconds) return '未设置';
    return this.formatExpiration(seconds);
  }
}

export const expirationService = new MessageExpirationService();
